pdp/expr_list_of_strings_test.go

```diff
@@ -212,6 +215,269 @@ func TestFunctionListOfStrings(t *testing.T) {
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("List of Strings Contains %v + %v", tc.a, tc.b), func(t *testing.T) {
+			a := MakeListOfStringsValue(tc.a)
+			b := MakeStringValue(tc.b)
+			e := makeFunctionListOfStringsContains(a, b)
+
+			v, err := e.Calculate(ctx)
+			if err != nil {
+				t.Errorf("Expect Calculate() returns no error, but got '%s'", err)
+				return
+			}
+
+			res, err := v.boolean()
+			if err != nil {
+				t.Errorf("Expect boolean result with no error, but got '%s'", err)
+			} else if res != tc.c {
+				t.Errorf("Expect result '%v', but got '%v'", tc.c, res)
+			}
+		})
+	}
+}
+
+func TestListOfStringsEqual(t *testing.T) {
+	ctx, err := NewContext(nil, 0, nil)
+	if err != nil {
+		t.Fatalf("Expected context but got error %s", err)
+	}
+
+	testCases := []struct {
+		a, b []string
+		c    bool
+	}{
+		{
+			a: []string{},
+			b: []string{},
+			c: true,
+		},
+		{
+			a: []string{"foo"},
+			b: []string{},
+			c: false,
+		},
+		{
+			a: []string{"foo"},
+			b: []string{"foo"},
+			c: true,
+		},
+		{
+			a: []string{"foo"},
+			b: []string{"bar"},
+			c: false,
+		},
+		{
+			a: []string{"foo", "bar"},
+			b: []string{"foo", "bar"},
+			c: true,
+		},
+		{
+			a: []string{"foo", "foo"},
+			b: []string{"foo"},
+			c: false,
+		},
+		{
+			a: []string{"foo", "bar"},
+			b: []string{"bar", "foo"},
+			c: false,
+		},
+		{
+			a: []string{"foo", "bar", "this"},
+			b: []string{"foo", "bar", "this", "extra"},
+			c: false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("List of Strings Equal %v + %v", tc.a, tc.b), func(t *testing.T) {
+			a := MakeListOfStringsValue(tc.a)
+			b := MakeListOfStringsValue(tc.b)
+			e := makeFunctionListOfStringsEqual(a, b)
+
+			v, err := e.Calculate(ctx)
+			if err != nil {
+				t.Errorf("Expect Calculate() returns no error, but got '%s'", err)
+				return
+			}
+
+			res, err := v.boolean()
+			if err != nil {
+				t.Errorf("Expect boolean result with no error, but got '%s'", err)
+			} else if res != tc.c {
+				t.Errorf("Expect result '%v', but got '%v'", tc.c, res)
+			}
+		})
+	}
+}
+
+func TestListOfStringsIntersect(t *testing.T) {
+	ctx, err := NewContext(nil, 0, nil)
+	if err != nil {
+		t.Fatalf("Expected context but got error %s", err)
+	}
+
+	testCases := []struct {
+		a, b, c []string
+	}{
+		{
+			a: []string{"foo", "bar", "doo"},
+			b: []string{"boo", "mar", "aoo"},
+			c: []string{},
+		},
+		{
+			a: []string{"foo", "bar"},
+			b: []string{"boo", "mar", "foo"},
+			c: []string{"foo"},
+		},
+		{
+			a: []string{"foo", "bar", "boo"},
+			b: []string{"boo", "mar", "foo"},
+			c: []string{"boo", "foo"},
+		},
+		{
+			a: []string{"foo", "foo", "bar", "bar", "bar"},
+			b: []string{"bar", "foo", "foo", "moo"},
+			c: []string{"bar", "foo", "foo"},
+		},
+		{
+			a: []string{"1", "2", "2", "3", "3", "3", "4", "4", "4", "4"},
+			b: []string{"4", "3", "3", "2", "2", "2", "1", "1", "1", "1"},
+			c: []string{"4", "3", "2", "1", "2", "3"},
```

Comment:
I did not envision this is how the list of strings intersect function would behave.

- Does this behavior satisfy your use case? Maybe you can provide a couple of examples?
- I was envisioning that it behaves similarly to set of strings, except that the result is not sorted.

Reply:
I have reviewed the use cases and I don't think number of repetitions is necessary to store. Changing the intersect algorithm to what you and Vasili recommended above will resolve this behavior.